### PR TITLE
Updating the table

### PR DESCRIPTION
**Pull request by MayPyone**
-Add a new column
-Insert new data
-Use aggregate functions 
-Use transactions

Screenshot (642)
Screenshot (643)
Screenshot (644)
Screenshot (645)
Screenshot (646)
Screenshot (647)
Screenshot (648)
Screenshot (650)
Screenshot (651)
Screenshot (652)

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -14,4 +14,30 @@ SELECT name,escape_attempts FROM animals WHERE weight_kg>10.5;
  /* all animals not named Gabumon. */
   SELECT * FROM animals WHERE NOT name='Gabumon';
   /* all animals with a weight between 10.4kg and 17.3kg */
-   SELECT * FROM animals WHERE weight_kg BETWEEN 10.4 AND 17.3;
\ No newline at end of file
+   SELECT * FROM animals WHERE weight_kg BETWEEN 10.4 AND 17.3;
+
+
+
+
```
Comment (topeogunleye) [inline queries.sql]: - [ ] Great job with your SQL transactions, however I kindly suggest that you implement the below project requirement that is missing from your queries.sql:

- Inside a transaction update the animals table by setting the species column to unspecified. Verify that change was made. Then roll back the change and verify that the species columns went back to the state before the transaction.

File: queries.sql [hunk 2]
```diff
@@ -14,4 +14,30 @@ SELECT name,escape_attempts FROM animals WHERE weight_kg>10.5;
...
+
+    update animals set species = 'digimon' where name LIKE '%mon%';
+    update animals set species = 'pokemon' where species='';
```
Comment (topeogunleye) [inline queries.sql:27]: Hello 👋,

Great job initiating the process of updating the `animals` table with the specified conditions! 🚀 Your SQL statements are on the right track in terms of updating species names based on the conditions provided in the project requirements. However, there's a crucial part about using a transaction that hasn’t been addressed in your submitted code snippet. 

1. **Using Transactions:**
   - **Requirement:** 📘 Ensure all SQL commands are executed within a transaction to maintain database integrity.
   - **Suggestion:** 🛠️ Use `BEGIN TRANSACTION;` to start, and `COMMIT;` to persist your changes.

2. **Verification Steps:**
   - **Requirement:** 📘 After executing your update commands, it’s necessary to verify the changes were made and persist after the commit.
   - **Suggestion** 🛠️  Use `SELECT` statements after the updates and the commit to ensure the data was updated as expected.

Here’s an example of how you might adjust your code:

```sql
   BEGIN;
   UPDATE animals SET species = 'digimon' WHERE name LIKE '%mon'; 
   UPDATE animals SET species = 'pokemon' WHERE species IS NULL; 

   SELECT species from animals; -- verify that change was made
   COMMIT;
   SELECT species from animals; -- verify that change persists after commit
```

Would love to see your progress as you incorporate these feedback points! If you have any questions or need further clarification on any of the above points, please feel free to ask.

Keep up the good work! 💻🚀

File: queries.sql [hunk 3]
```diff
@@ -14,4 +14,30 @@ SELECT name,escape_attempts FROM animals WHERE weight_kg>10.5;
...
+    delete from animals where date_of_birth > '2022-01-01';
+    /*  savepoint s1; */
+    update animals set weight_kg = weight_kg * -1;
+    /* ROLLBACK to s1; */
+    update animals set weight_kg = weight_kg * -1 where weight_kg<0;
+    /* commit; */
```
Comment (topeogunleye) [inline queries.sql]: Hello! 👋🏾

 I can see the effort and thought that you've put into trying to implement the SQL operations, and it’s super commendable! 🏆💻 Your starting point provides a solid basis, and with a bit of tweaking, we'll have it fully aligned with the project requirements. 🎉🛠️

```sql
delete from animals where date_of_birth > '2022-01-01';   
delete from animals where date_of_birth > '2022-01-01';
/* savepoint s1; */
update animals set weight_kg = weight_kg * -1;
/* ROLLBACK to s1; */
update animals set weight_kg = weight_kg * -1 where weight_kg<0;
/* commit; */
```
It's great that you've made a start! 🚀 However, it seems like there are some aspects to adjust to fully meet the project requirements. Let’s navigate through them together! 🤝

**Project Requirements Quick Peek:** 🧐📜
1. Delete all animals born after Jan 1st, 2022 🗑️.
2. Create a savepoint for the transaction 📍.
3. Update all animals' weight to be their weight multiplied by -1 🔄.
4. Rollback to the savepoint 🔙.
5. Update all animals' weights that are negative to be their weight multiplied by -1 ⚖️.
6. Commit the transaction ✔️.

Here's a breakdown of your code compared to the requirements:

1. **Deleting Records**: ✅ Your delete statement aligns with the requirements!
```sql
delete from animals where date_of_birth > '2022-01-01';  
```
2. **Utilizing Transaction and Savepoint**: ❌ This part needs a bit of adjustment. Ensure to utilize the `BEGIN TRANSACTION` and create a savepoint after the deletion.
3. **Updating Weights**: ✅ The concept for updating weight is present, though we need to make sure this occurs within our transaction.
4. **Rollback & Update**: ❌ Ensure to actually rollback to the savepoint and perform updates within the transaction scope.
5. **Commit**: ❌ The commit statement should be uncommented to persist the changes.

Here's a skeleton to guide you through crafting a compliant solution 🚀:
```sql
BEGIN TRANSACTION;

-- Your DELETE operation here

SAVEPOINT your_savepoint_name;

-- Your UPDATE operation here

ROLLBACK TO your_savepoint_name;

-- Your second UPDATE operation here

COMMIT;
```
Please feel free to explore, revise, and ask anything - I’m here for all your questions and to cheer you on! 🤗🎈

File: queries.sql [hunk 4]
```diff
@@ -14,4 +14,30 @@ SELECT name,escape_attempts FROM animals WHERE weight_kg>10.5;
...
   SELECT * FROM animals WHERE NOT name='Gabumon';
   /* all animals with a weight between 10.4kg and 17.3kg */
-   SELECT * FROM animals WHERE weight_kg BETWEEN 10.4 AND 17.3;
\ No newline at end of file
+   SELECT * FROM animals WHERE weight_kg BETWEEN 10.4 AND 17.3;
+
+
+
+
+
```
Comment (topeogunleye) [inline queries.sql]: - [ ] Great job on implementing these SQL project requirements however the below project requirement hasn't been implemented yet:

Project Requirements Sneak Peak: 🧐📜

- Inside a transaction delete all records in the animals table, then roll back the transaction.
- After, rollback verify if all records in the animals table still exists. 

Please feel free to explore, revise, and ask anything - I’m here for all your questions and to cheer you on! 🤗🎈

File: queries.sql [hunk 5]
```diff
@@ -14,4 +14,30 @@ SELECT name,escape_attempts FROM animals WHERE weight_kg>10.5;
...
+    update animals set weight_kg = weight_kg * -1 where weight_kg<0;
+    /* commit; */
+  
+   /* total animals */
+    select count(id) from animals;
+    /* animals that have never tried to escape */
+    select count(id) from animals where escape_attempts = 0;
+    /* the average weight of animals */
+     select max(escape_attempts) from animals;
+     select name, neutered from animals where escape_attempts = 7;
```
Comment (topeogunleye) [inline queries.sql]: Hello there! 🌟

First and foremost, **huge kudos** to you for putting in the effort and submitting your SQL code! 🎉💻 Your willingness to tackle the task and provide a solution is evident, and that’s something truly worth celebrating! 🥳👏 

Let’s delve into the specifics and get you even closer to achieving the desired outcomes! 🚀📘

**Quick Recap of Project Requirements:** 🧐📜
- Find the average weight of animals.

**Snapshot of Your Code:** 🖥️
```sql
select max(escape_attempts) from animals;
select name, neutered from animals where escape_attempts = 7;
```
While your submitted code displays a solid grasp on SQL querying, it seems we've veered off course just a tad with regards to the specific project requirements! 🧭📏 No worries at all - we're all here to learn and grow! 🌱📈 

Here’s some tailored feedback to guide your revision: 🧐✨
- Your queries seem to be centered around `escape_attempts`, whereas our mission is to delve into animal weights. 🐾⚖️
  
**Guidance Towards the Goal:** 🚀🗺️
```sql
SELECT AVG(weight_kg) FROM animals;
```
This one-liner query directly addresses our requirement of fetching the average weight of animals from our dataset! 📊🐾

**A Friendly Tip:** 💡 Always make sure to thoroughly re-read the project requirements before diving into coding. Aligning your thought process with the objectives will set you on a straight path towards accurate and efficient solutions! 🎯🧠

**Keep Going!** 🎉💪 The journey of coding comes with its twists and turns, and every attempt is a step towards becoming a SQL wizard! 🧙‍♂️💼 Your forward momentum is palpable, and with each iteration, you're unfolding more and more of your potential! 🌟🛤️

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @MayPyone, 

Good job so far!, Amazing SQL Explorer! 🚀🔍,
Big applause to you for diving deep into the SQL oceans and retrieving pearls of wisdom! 🌊🐚 Your contributions to both schema.sql and data.sql were on-point and followed the project guidelines beautifully. 

Try-Again

### Highlight Review 🌟
#### Triumph in `schema.sql` 🏛️✨
- **Exquisite Addition to Schema:** 🧱📜 Your addition of a new `species` column with appropriate datatype and size is impeccable! Your choice of `varchar(255)` caters well to our use-case, providing ample space for varied species names while keeping data storage in check. 🌈🐾
```sql
ALTER TABLE animals ADD species varchar(255);
```

#### Mastery in `data.sql` 📊🌸
- **Perfect Data Insertions:** 🍀🚀 You have successfully inserted the given data into the `animals` table! Your meticulousness in adhering to the provided data and ensuring each value is in its rightful place demonstrates a keen eye for detail and a robust understanding of `INSERT` operations.
```sql
INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
VALUES ('Charmander', '2020/2/8', '0', 'false', '-11.0'),
       ('Plantmon', '2021/11/15', '2', 'true', '-5.7'),
       ('Squirtle', '1993/04/02', '3', 'false', '-12.13'),
       ('Angemon', '2005/06/12', '1', 'true', '-45.0'),
       ('Boarmon', '2005/06/07', '7', 'true', '20.4'),
       ('Blossom', '1998/10/13', '3', 'true', '17.0'),
       ('Ditto', '2022/05/14', '4', 'true', '22.0');
```
- **Efficient Batch Insertion:** 🚅📬 Your usage of a single `INSERT INTO` statement to populate multiple records demonstrates an understanding of efficient querying and effective database interaction. This is a noteworthy practice that minimizes server requests and optimizes transaction time!

#### Encouraging Words 🎉💖
Your progression and application of SQL commands in this project have been superb! 🚀🌟 By perfectly aligning your code with the project requirements, you’ve illustrated a beautiful synthesis of comprehension and application in the SQL universe! 🌌🗝️

Always remember: Your journey in the technological world is not just about acquiring knowledge, but also about celebrating every small victory along the way - just like today! 🌟🚀 Keep the curiosity alive, continue exploring, and never forget to bask in the joy of your accomplishments! 🌈🏆

Cheers to your amazing work and to the many more successes that await you on your learning adventure! 🎊🚀

### End of Highlight Review 🎈📚

Ready to sail towards new horizons? Let’s continue navigating through the sea of knowledge together! 🌊🚢📘

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- Nothing to mention.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql [hunk 6]
```diff
@@ -16,25 +16,41 @@ SELECT name,escape_attempts FROM animals WHERE weight_kg>10.5;
...
+    BEGIN TRANSACTION;
+    delete from animals;
+    select * from animals;
+    ROLLBACK;
+    select * from animals;
+
+    BEGIN TRANSACTION;
     delete from animals where date_of_birth > '2022-01-01';
-    /*  savepoint s1; */
+    savepoint s1;
```
Comment (TadesseDev) [inline queries.sql:40]: - [Optional] Good job creating a save point @MayPyone, But I would suggest having a meaning full name while doing so. This helps to make your code more readable and communicate what the save point is for.

**Review by TadesseDev — APPROVED** [review]
### Hi @MayPyone 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
AmazingCreepyFrogmouth-max-1mb




## Highlights
 - [x] Good job making the changes requested by the previous code reviewer. ✅


## Optional suggestions
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**